src/aarch64/internal.h

```diff
@@ -66,3 +66,24 @@ SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.  */
 #define N_X_ARG_REG		8
 #define N_V_ARG_REG		8
 #define CALL_CONTEXT_SIZE	(N_V_ARG_REG * 16 + N_X_ARG_REG * 8)
+
+/* Helpers for writing assembly compatible with arm ptr auth */
+#ifdef LIBFFI_ASM
+
+#ifdef HAVE_PTRAUTH
```

Comment:
this seems to be for generic code not just for apple which this is wrong:

- return address signing requires cfi annotation (toggle pac state)
- b key requires cfi annotation
- libffi should use nop-space instructions. armv8.3-a only asm can be added conditionally, but that's not relevant to most users: no linux distro would deploy binaries with armv8.3-a baseline any time soon.
- pauth and bti should go together in gcc target libraries: branch-protection requires both return address and indirect jump/call protection. at least that's the intended usage model. however bti support requires significant changes in the jump table as bti j does not fit there.

so if apple requires this then it should be apple only or changed such that it works for other targets too (at least aarch64*-linux-gnu), libitm aarch64/sjlj.S is an example how this was done in other target libraries.